File: cpp/day13/day13.cpp
#include <iostream>
#include <chrono>
#include "../utils/AdventReader.h"
#include "../utils/Maths.h"

#include <algorithm>

#include <map>

using namespace std;
using namespace avreader;


vector<vector<char>> buildGrid(const vector<string> &group) {
    vector<vector<char>> grid;

    for (auto line : group) {
        vector<char> tmp;
        for (auto c : line) {
            tmp.push_back(c);
        }
        grid.push_back(tmp);
    }

    return grid;
}

void checkReflections(const vector<vector<char>> &grid, int &p1, int &p2) {

    for (int row = 0; row < grid.size(); row++) {
        int rowRange = std::min(row, int(grid.size() - row));
        int upRow = row - 1;
        int downRow = row;

        int matchCount = 0;

        for (int dRow = 0; dRow < rowRange; dRow++) {
            auto upLine = grid[upRow - dRow];
            auto downLine = grid[downRow + dRow];

            for (int index = 0; index < upLine.size(); index++) {
                if (upLine[index] == downLine[index])
                {
                    matchCount++;
                }
            }
        }

        auto perfectMatch = rowRange * grid[row].size();
        if (matchCount == perfectMatch)
        {
            p1 = row;
        }
        else if (matchCount == perfectMatch - 1)
        {
            p2 = row;
        }
        
    }
}

int main() {


    runAndPrintMicros([]() {
        vector<vector<string>> groups = parseFileToGroups("day13.txt", "");

        vector<vector<string>> grid;

        long p1,p2;
        for (auto group : groups) {

            auto grid = buildGrid(group);
            
            int ver1, ver2 = 0;
            checkReflections(grid, ver1, ver2);

            auto transposed = maths::transpose(grid);
            int hor1, hor2 = 0;
            checkReflections(transposed, hor1, hor2);

            p1 += ver1 * 100 + hor1;
            p2 += ver2 * 100 + hor2;
        }


        cout << "Part 1 = " << p1 << endl;//37975
        cout << "Part 2 = " << p2 << endl;//32497

    });//4831, 7131, 4785

    return 0;
}

File: 2023/go/day13/day13.go
package main

import (
	"advent/lib"
	"advent/lib/maths"
	"fmt"
)

func main() {

	lib.RunAndPrintDuration(func() {
		groups := lib.ReadFileToGroups("day13.txt", "")

		var p1, p2 int64 = 0, 0
		for i,group := range groups {
			fmt.Println("Group ", i)
			grid := buildGrid(group)

			ver1, ver2 := checkReflect(grid);

			transposed := maths.Transpose[rune](grid)
			hor, hor2 := checkReflect(transposed);

			p1 += ver1*100 + hor
			p2 += ver2*100 + hor2
		}

		fmt.Println("Part 1 = ", p1)//37975
		fmt.Println("Part 2 = ", p2)//32497
	})//780, 734, 782, 729
}

func buildGrid(group []string) [][]rune {
	var grid [][]rune

	for _, row := range group {
		temp := make([]rune, 0)
		for _, c := range row {
			temp = append(temp, c)
		}
		grid = append(grid, temp)
	}

	return grid
}

func checkReflect(group [][]rune) (int64, int64) {
	var p1, p2 int64


	//For each row 
	for row := 1; row < len(group); row++ {
		reflectRange := min(row, len(group) - row)//The range to search for reflection
		upRow, downRow := row - 1, row

		matchCount := 0

		//For each possible row in reflection
		//Compare chars in each row, increment count if found
		for dRow := 0; dRow < reflectRange; dRow++ {
			upLine, downLine := group[upRow - dRow], group[downRow + dRow]

			for index := 0; index < len(upLine); index++ {
				if downLine[index] == upLine[index] {
					matchCount++
				}
			}
		}

		//Maximum possible char matches is number of rows compared times length of rows
		perfectCount := reflectRange * len(group[row])

		if matchCount == perfectCount {
			p1 = int64(row)
		} else if matchCount == perfectCount - 1 {
			p2 = int64(row)
		}

	}


	return p1, p2
}